src/adapt.ts

```diff
@@ -198,18 +198,31 @@ export function adaptOutput(rawResult: string, outputSpecs: starknet.Argument[],
             adapted[outputSpec.name] = currentValue;
             resultIndex++;
 
-        } else if (outputSpec.type === "felt*") {
+        } else if (outputSpec.type.endsWith("*")) {
             const lenName = `${outputSpec.name}${LEN_SUFFIX}`;
             if (lastSpec.name !== lenName || lastSpec.type !== "felt") {
-                const msg = `Array size argument ${lenName} (felt) must appear right before ${outputSpec.name} (felt*).`;
+                const msg = `Array size argument ${lenName} (felt) must appear right before ${outputSpec.name} (${outputSpec.type}).`;
                 throw new HardhatPluginError(PLUGIN_NAME, msg);
             }
 
+            // Remove * from the spec type
+            const outputSpecArrayElementType = outputSpec.type.slice(0, -1);
             const arrLength = Number(adapted[lenName]);
-            const arr = result.slice(resultIndex, resultIndex + arrLength);
-            adapted[outputSpec.name] = arr;
-            resultIndex += arrLength;
 
+            const structArray = [];
+
+            // Iterate over the struct array, starting at index, starting at `resultIndex`
+            let newRawIndex = resultIndex;
```

Comment:
no need for `newRawIndex`, just passing and assigning to the same `resultIndex` should be enough